src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -74,68 +87,65 @@ CoreWorkerDirectTaskSubmitter::GetOrConnectLeaseClient(
 }
 
 void CoreWorkerDirectTaskSubmitter::RequestNewWorkerIfNeeded(
-    const TaskSpecification &resource_spec, const rpc::Address *raylet_address) {
-  if (worker_request_pending_) {
+    const SchedulingClass &scheduling_class, const rpc::Address *raylet_address) {
+  if (pending_lease_requests_.find(scheduling_class) != pending_lease_requests_.end()) {
     return;
   }
-  if (queued_tasks_.empty()) {
+  auto it = task_queues_.find(scheduling_class);
+  if (it == task_queues_.end()) {
     // We don't have any tasks to run, so no need to request a worker.
     return;
   }
 
-  // NOTE(swang): We must copy the resource spec here because the resource spec
-  // may get swapped out by the time the callback fires. If we change this so
-  // that we associate the granted worker with the requested resource spec,
-  // then we can just pass the ref instead of copying.
-  TaskSpecification resource_spec_copy(resource_spec.GetMessage());
   auto lease_client = GetOrConnectLeaseClient(raylet_address);
+  TaskSpecification &resource_spec = it->second.front();
+  TaskID task_id = resource_spec.TaskId();
   RAY_CHECK_OK(lease_client->RequestWorkerLease(
-      resource_spec_copy,
-      [this, resource_spec_copy, lease_client](
+      resource_spec,
+      [this, lease_client, task_id, scheduling_class](
           const Status &status, const rpc::WorkerLeaseReply &reply) mutable {
         if (status.ok()) {
           if (!reply.worker_address().raylet_id().empty()) {
-            RAY_LOG(DEBUG) << "Lease granted " << resource_spec_copy.TaskId();
+            RAY_LOG(DEBUG) << "Lease granted " << task_id;
             HandleWorkerLeaseGranted(
```

Comment:
If you don't mind doing it in this PR, we could clean this code up a bit by moving the `absl::MutexLock` and `pending_lease_requests_.erase(...)` to the beginning of the callback (I think we currently do this in all branches except the RAY_LOG(FATAL) one).